src/Dfc.CourseDirectory.FindACourseApi/Features/TLevels/TLevels.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Dfc.CourseDirectory.Core.DataStore.CosmosDb;
+using Dfc.CourseDirectory.Core.DataStore.Sql;
+using Dfc.CourseDirectory.Core.DataStore.Sql.Queries;
+using MediatR;
+using CosmosQueries = Dfc.CourseDirectory.Core.DataStore.CosmosDb.Queries;
+
+namespace Dfc.CourseDirectory.FindACourseApi.Features.TLevels
+{
+    public class Query : IRequest<ViewModel>
+    {
+    }
+
+    public class ViewModel
+    {
+        public IReadOnlyCollection<TLevelDetailViewModel> TLevels { get; set; }
+    }
+
+    public class Handler : IRequestHandler<Query, ViewModel>
+    {
+        private readonly ISqlQueryDispatcher _sqlQueryDispatcher;
+        private readonly ICosmosDbQueryDispatcher _cosmosDbQueryDispatcher;
+
+        public Handler(ISqlQueryDispatcher sqlQueryDispatcher, ICosmosDbQueryDispatcher cosmosDbQueryDispatcher)
+        {
+            _sqlQueryDispatcher = sqlQueryDispatcher;
+            _cosmosDbQueryDispatcher = cosmosDbQueryDispatcher;
+        }
+
+        public async Task<ViewModel> Handle(Query request, CancellationToken cancellationToken)
+        {
+            var tLevels = await _sqlQueryDispatcher.ExecuteQuery(new GetTLevels());
+
+            var providerIds = tLevels.Select(t => t.ProviderId).Distinct();
+
+            var getProviders = _cosmosDbQueryDispatcher.ExecuteQuery(
+                new CosmosQueries.GetProvidersByIds { ProviderIds = providerIds });
+
+            var getSqlProviders = _sqlQueryDispatcher.ExecuteQuery(
+                new GetProvidersByIds { ProviderIds = providerIds });
+
+            var getVenues = _sqlQueryDispatcher.ExecuteQuery(
+                new GetVenuesByIds() { VenueIds = tLevels.SelectMany(t => t.Locations.Select(l => l.VenueId)).Distinct() });
+
+            await Task.WhenAll(getProviders, getSqlProviders, getVenues);
+
+            var providers = await getProviders;
+            var sqlProviders = await getSqlProviders;
+            var venues = await getVenues;
+
+            var feChoices = await _cosmosDbQueryDispatcher.ExecuteQuery(new CosmosQueries.GetFeChoicesByProviderUkprns { ProviderUkprns = providers.Values.Select(p => p.Ukprn) });
+
+            return new ViewModel
+            {
+                TLevels = tLevels.Select(t =>
+                {
+                    var provider = providers[t.ProviderId];
```

Comment:
Pondering: Should this blow up if we have invalid data? Or should we continue to serve up what valid data we have?

Reply:
Seems fine as-is.